### ISSUE TITLE
need RFC for stdio representation in KVS

### ISSUE THREAD

**Issue by garlick** [flux-framework/rfc#192]
The stdin, stdout, and stderr of a job should be part of the job's permanent record in the KVS by default.  Since output may be parsed long after it was written, the representation should be documented.  The representation should have the following characteristics:
* The KVS value is a sequence of chunks
* Chunks may be appended, with intermediate values still a valid sequence
* A chunk may contain binary data (0x00 - 0xff)
* Chunk metadata includes _timestamp_, _task rank_, and _stream_ id

**Comment by garlick** [comment]
Pondering this a bit this morning, I was reminded of flux-framework/flux-core#932 (use UTF-8 encoding), which I think we dismissed as wrong headed due to the fact that this would communicate that the output is "text" when it might not be.  (Sadly the issue wasn't updated after we talked about it, so maybe I've forgotten where that ended up)

Should an encoding include the equivalent of a `content-type` field, with a default of "text" and a user-controlled option for "octet-stream" or similar?  That would let us represent text directly as a JSON string, including UTF-8 code points if used.  Avoiding base64 for text would be more efficient and less obfuscated for future alien decoding.  Yet if a user wants to use stdio for non-textual data, they could do that by providing an option to the shell, and then aliens would know whether the bytes are human readable or not...

For actual implementation, the shell will encounter a jansson error if it tries to encode invalid UTF-8 as a string.  Would this be fatal to the job?

This implies a "header" portion in the KVS representation.  Hmm, even if I'm totally off base with imposing an interpretation on the data, maybe a header is still a good idea, as it would allow us to version the encoding without repeating it in each blob...

**Comment by grondo** [comment]
Yeah, that seems to make sense. Realistically, we can't expect users to provide the content-type of their stdio for jobs, and we'd have to protect against jobs inadvertently sending binary data as output as you hinted above (the equivalent of accidentally cat-ing a binary file to your terminal). However, the pty option could probably turn on the correct encoding, which is the main case we're worried about now.

I guess the main question is can we get it done in the time allotted for this work? If not, then we have to decide if the benefit is worth pushing back the deadline. Whereas, we already have several working base64 based solutions since it is simple, and we could update to the superior solution down the road after we have something working.

**Comment by garlick** [comment]
Yeah, maybe for expediency, we could use the existing message format(s) or a refactoring to a common format as discussed, and then for the KVS representation (the subject of this issue), use the [RFC 18](https://github.com/flux-framework/rfc/blob/master/spec_18.adoc) KVS eventlog format?  It already has the property of remaining valid across appends, has a timestamp per entry, and we could use the event name for an initial `header` event, followed by multiple `stdout`, `stdin`, `stderr` events, and the context for header-indicated data encoding?

**Comment by garlick** [comment]
E.g. header event context could look like:
```json
{
  "timestamp":1552593348.073045,
  "name":"io-header",
  "context":{
    "version":1,
    "stdin-encoding":"base64",
    "stdout-encoding":"base64",
    "stderr-encoding":"UTF-8"}
  }
}
```

**Comment by grondo** [comment]
>  and then for the KVS representation (the subject of this issue), use the RFC 18 KVS eventlog format?

I *really* like this idea! I even like the co-mingling of stdin/out/err, which I had never considered, seems to me to be a simplifying improvement.

The eventlog could also be used by other I/O modes/plugins to redirect I/O away from the kvs if desired, e.g. a mode where output goes only to files could generate an event with the full path to any file(s) opened for output...

**Comment by garlick** [comment]
To continue the discussion, here are some  straw man events for the stdio eventlog:

Header:
```json
{
  "timestamp":1552593348.073045,
  "name":"io-header",
  "context":{
    "version":1,
    "stdin-encoding":"base64",
    "stdout-encoding":"base64",
    "stderr-encoding":"UTF-8",
    "task-count": 32,
    "options":[
        "stdin-bcast"
    ]
  }
}
```

Stdout from rank 31 with base64 encoding (rank is a string in case we want an idset there):
```json
{
  "timestamp":1552593349.1,
  "name":"stdout",
  "context":{
    "rank":"31",
    "data":"aGVsbG8gd29ybGQK",
  }
}
```
Stderr from rank 0 with UTF-8 encoding:
```json
{
  "timestamp":1552593349.2,
  "name":"stderr",
  "context":{
    "rank":"0",
    "data":"Error: something failed\n",
  }
}
```

Stdin with base64 encoding:
```json
{
  "timestamp":1552593349.4,
  "name":"stdin",
  "context":{
    "data":"bWVlcAo=",
  }
}
```

EOF for stdin:
```json
{
  "timestamp":1552593349.5,
  "name":"stdin",
  "context":{
    "eof":true
  }
}
```

EOF for stdout (here combined for all tasks, but could be per-task if requested?):
```json
{
  "timestamp":1552593349.5,
  "name":"stdout",
  "context":{
    "rank":"0-31",
    "eof":true
  }
}
```

Footer (always the last event):
```json
{
  "timestamp":1552593349.6,
  "name":"io-summary",
  "context":{
    "stdin-bytes":5,
    "stdout-bytes":1024,
    "stderr-bytes":24
}
```

**Comment by grondo** [comment]
Great!

Some quick comments:

 * We may want to support more than one stdin stream, and also specify which tasks get stdin. E.g. a common use case would be to send stdin to rank 0 and close all other tasks input immediately. We could just add a `rank` key like the context of the `stdout` `stderr` events, but there are probably other ways to do it as well (e.g. I see you considered handling this via configuration in the header event). Another idea would be to have named input streams and configure which tasks read from which stream.

 * The options in the io-header should probably be an object not an array of strings, for flexibility.

 * We should allow `data` and `eof` to appear in the same event (not explicitly disallowed here), since this could reduce the number of required events, especially with batching.

 * For the io-header, it would be slightly more compact to keep the encoding configuration under a single key, e.g.: `"encoding": { "stdout": "base64", "stdin":base64", ...}`. This also allows the encoding config to be slightly more extensible.

 * Do we need a way to different encoding per-task? e.g. If rank 0 is using a pty, it would possibly need a different encoding than other tasks.

Finally, we might want to allow a way for other parts of the job execution system to write to the job's output. E.g. in wreck it was very handy to have an error log which the frontend tool could display on stderr. With this scheme, any part of the execution system could actually write directly to the job's stderr by appending a `stderr` event, however the `rank` key may not make much sense in that context. Perhaps another type of event would be useful here?

**Comment by garlick** [comment]
Closed by #194
